keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +337,302 @@ or later. The dual-stack feature requires no changes to this API.
 The versions of CNI plugin binaries that must be used for proper dual-stack functionality (and IPv6 functionality in general) depend upon the version of the container runtime that is used in the cluster nodes (see [CNI issue #531](https://github.com/containernetworking/cni/issues/531) and [CNI plugins PR #113](https://github.com/containernetworking/plugins/pull/113)):
 
 ### Services
+Services depend on `--service-cluster-ip-range` for `ClusterIP` assignment. A dualstack cluster can have this flag configured as `--services-cluster-ip-rang=<cidr>,<cidr>` or `--services-cluster-ip-rang=<cidr>`. The following rules apply:
+
+1. The maximum allowed number of cidrs in this flag is two. Attempting to use more than two cidrs will generate failures during startup. 
+2. When configured with two cidrs then the values are expected to have different ip families. e.g `--service-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>` or `--service-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`. Attempting to configure a cluster with two similar cidrs will generate failures during startup.
+3. It is possible to have a `partially-dual-stack` cluster, with dual-stack pod networking but only a single service cidr. This means that pods in the cluster will be able to send and receive traffic on both ip families, but services which use the `clusterIP` or `ClusterIPs` field will only be able to use the ip family of this flag.
+4. When upgrading a cluster to dualstack, the first element of the `--service-cluster-ip-range` flag MUST NOT change IP families.  As an example, an IPv4 cluster can safely be upgraded to use `--service-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>`, but can not safely be changed to `--service-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`.
+5. The first cidr entry of `--service-cluster-ip-range=<first cidr entry>, <secondary cidr entry>` is considered cluster default ip address family.
+6. Attempting to allocate `ClusterIP` or `ClusterIPs` from an ip address family that is not configured on the cluster will result in an error.
+
+> The below discussion assumes that `Endpoint Slice and Endpoint Slice Controller`  will allow dual stack endpoints, the existing endpoint controller will remain single stack. That means enabling dual stack services by having more than one entry in `--service-cluster-ip-range` without enabling `Endpoint Slice` feature gate will result in validation failures during startup.
```

Comment:
OK, but even when using EndpointSliceController, we still also run the EndpointController and generate Endpoints for compatibility with old components, correct? So we still need to specify how EndpointController behaves in a dual-stack cluster... (Actually, we need to specify how EndpointSliceController behaves in a dual-stack cluster too... you didn't update that part of the KEP...)

Reply:
@robscott Was weighing options, but I think we have a canonical answer unless we choose to change it

Reply:
So it sounds like this is suggesting Endpoints resources would be single stack only. As far as the Endpoints controller, it would be relatively straightforward to drop any IPs not matching the family of the Endpoints resource. 

If we want to say that the Endpoints resource is guaranteed to be single stack, that may be more difficult. Of course the Endpoints controller is not the only thing that can produce/manage Endpoints. There is a chance that someone is already writing dual-stack Endpoints in a cluster somewhere. Would we need to add some kind of field and/or validation to enforce this single stack behavior? Maybe it's fine to only update the behavior of the controller here and leave the API alone?

For the EndpointSlice controller I think the answer is somewhat simpler. Assuming we maintain the IPv4 and IPv6 address types, the EndpointSlice controller could just be modified to detect the address type and add IP addresses to EndpointSlices with the appropriate address type. So if all Pods backing a Service have both v4 and v6 IPs, we'd end up with 2x the number of EndpointSlices for a dual stack Service, half for the v4 addresses and the other half for the v6 addresses.

Reply:
@robscott yeah, let's be clear - when an `Endpoints` is managed by the Enpoints controller, it will be single-stack, corresponding to `ipFamilies[0]`.  `Endpoints` managed by anything else is out of scope.

For EPSlice - I'm still looking to you for thoughts on whether you think merging back to a single type (IP) or keeping distinct IPv4, IPv6 makes sense.  Or both, which gives creators freedom, but forces consumers to have to handle both.  Once you GA, you have to keep whatever you choose.